server/src/main/java/io/deephaven/server/arrow/ArrowFlightUtil.java

```diff
@@ -69,6 +71,185 @@ public class ArrowFlightUtil {
+                    msg.modColumnData = ZERO_MOD_COLUMNS; // no mod column data for DoGet
+                    long elapsed = System.nanoTime() - start;
+                    // accumulate snapshot time in the metrics
+                    metrics.snapshotNanos += elapsed;
+
+                    // send out the data. Note that although a `BarrageUpdateMetaData` object will
+                    // be provided with each unique snapshot, vanilla Flight clients will ignore
+                    // these and see only an incoming stream of batches
+                    try (final BarrageStreamGenerator bsg = new BarrageStreamGenerator(msg, metrics)) {
+                        if (rsIt.hasMore()) {
+                            listener.onNext(bsg.getSnapshotView(snapshotRequestOptions,
+                                    snapshotViewport, false,
+                                    msg.rowsIncluded, columns));
+                        } else {
+                            listener.onNext(bsg.getSnapshotView(snapshotRequestOptions,
+                                    viewport, reverseViewport,
+                                    msg.rowsIncluded, columns));
+                        }
+                    }
+
+                    if (msg.rowsIncluded.size() > 0) {
+                        // very simplistic logic to take the last snapshot and extrapolate max
+                        // number of rows that will not exceed the target UGP processing time
+                        // percentage
+                        long targetNanos = (long) (TARGET_SNAPSHOT_PERCENTAGE
+                                * UpdateGraphProcessor.DEFAULT.getTargetCycleDurationMillis()
+                                * 1000000);
+
+                        long nanosPerCell = elapsed / (msg.rowsIncluded.size() * columnCount);
+
+                        // apply an exponential moving average to filter the data
+                        if (snapshotNanosPerCell == 0) {
+                            snapshotNanosPerCell = nanosPerCell; // initialize to first value
+                        } else {
+                            // EMA smoothing factor is 0.1 (N = 10)
+                            snapshotNanosPerCell =
+                                    (snapshotNanosPerCell * 0.9) + (nanosPerCell * 0.1);
+                        }
+
+                        snapshotTargetCellCount =
+                                (long) (targetNanos / Math.max(1, snapshotNanosPerCell));
+                    }
+                }
+            }
+        }
+    }
+
+    private static void createAndSendSnapshot(BaseTable table, BitSet columns, RowSet viewport, boolean reverseViewport,
+            BarrageSnapshotOptions snapshotRequestOptions, StreamObserver<BarrageStreamGenerator.View> listener,
+            BarragePerformanceLog.SnapshotMetricsHelper metrics) {
+
+        // if the table is static or append-only and a full snapshot is requested, we can make and send multiple
```

Comment:
I believe the `append-only` portion of this comment is invalid - as in we don't have a way to detect this.